update-versions.py

```diff
@@ -60,58 +100,83 @@ def calculate_hash(versions, version):
-    if current_vendor_hash:
+def calculate_vendor_hash(
+    version: semver.Version,
+    maybe_current_hashes: Optional[NixHashes],
+    calculated_hash: str,
+    vendor_hash_nix: Path,
+) -> str:
+    if maybe_current_hashes:
         print(f"Using existing vendorHash for {version}")
-        return current_vendor_hash
+        return maybe_current_hashes.vendorHash
     else:
         print(f"Calculating vendorHash for {version}")
         return nix_prefetch(
             [
                 "--file",
-                str(vendor_hash.resolve()),
+                str(vendor_hash_nix.resolve()),
                 "--argstr",
                 "version",
-                version,
+                str(version),
                 "--argstr",
                 "hash",
                 calculated_hash,
             ]
         )
 
 
-def nix_prefetch(args):
+def nix_prefetch(args: Iterable[str]) -> str:
     return subprocess.check_output(
         [
             "nix-prefetch",
             "--silent",
             "--option",
             "extra-experimental-features",
             "flakes",
-        ]
-        + args,
+            *args,
+        ],
         text=True,
     ).strip()
 
 
-parser = argparse.ArgumentParser(description="Update versions.json file")
-parser.add_argument("--vendor_hash", type=pathlib.Path, default="vendor-hash.nix")
-args = parser.parse_args()
-
-auth = github.Auth.Token(os.environ["GITHUB_TOKEN"])
-g = github.Github(auth=auth)
-repo = g.get_repo("hashicorp/terraform")
-# TODO: Drop "v" prefix first
-releases = list(filter(is_stable, repo.get_releases()))
-current_versions = read_versions()
-versions = collections.OrderedDict(
-    sorted(
-        functools.reduce(
-            to_version(args.vendor_hash), releases, current_versions
-        ).items(),
-        reverse=True,
+def main():
+    parser = argparse.ArgumentParser(description="Update versions.json file")
+    parser.add_argument(
+        "--vendor_hash",
+        type=pathlib.Path,
+        default="vendor-hash.nix",
+        help="Path to vendor-hash.nix file",
     )
-)
-with open("versions.json", "w") as f:
-    json.dump(versions, f, indent=2)
+    args = parser.parse_args()
+
+    gh_token = github.Auth.Token(os.environ["GITHUB_TOKEN"])
+    gh = github.Github(auth=gh_token)
+    repo = gh.get_repo("hashicorp/terraform")
+
+    versions_file = Path("versions.json")
+
+    gh_releases = repo.get_releases()
+    stable_gh_versions = get_stable_github_versions(gh_releases)
+    current_versions = read_current_versions(versions_file)
+
+    versions = collections.OrderedDict(
+        sorted(
+            functools.reduce(
+                get_or_calculate_hashes(args.vendor_hash),
+                stable_gh_versions,
+                current_versions,
+            ).items(),
+            reverse=True,
+        )
+    )
+
+    versions_jsonified = OrderedDict(
+        (str(version), hashes.__dict__) for version, hashes in versions.items()
+    )
```

Comment:
This hack is required because `json.dump` allows only `str` as keys. So we have to create another `OrderedDict` (it preserves  insert order) with string keys.

Reply:
@pedorich-n I think that is fine for the time being.